File: src/LangDeps.java
// LangBase is a language-dependent code used in GreenTea.java

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;

public abstract class LangDeps {

	public final static void println(String msg) {
		System.out.println(msg);		
	}

	public final static void DebugP(String msg) {
		if(GtStatic.DebugPrintOption) {
			LangDeps.println("DEBUG" + LangDeps.GetStackInfo(2) + ": " + msg);
		}
	}

	public final static String GetStackInfo(int depth){
		String LineNumber = " ";
		try{
			throw new Exception();
		}
		catch(Exception e){
			StackTraceElement[] Elements = e.getStackTrace();
			if(depth < Elements.length){
				StackTraceElement elem = Elements[depth];
				LineNumber += elem;
			}
		}
		return LineNumber;
	}

	public final static void Exit(int status, String Message) {
		System.err.println(Message);
		System.exit(1);
		//throw new RuntimeException(Message);
	}

	public final static void Assert(boolean TestResult) {
		if(!TestResult) {
			Exit(1, "Assertion Failed");
		}
	}
	
	public final static boolean IsWhitespace(char ch) {
		return Character.isWhitespace(ch);
	}

	public final static boolean IsLetter(char ch) {
		return Character.isLetter(ch);
	}

	public final static boolean IsDigit(char ch) {
		return Character.isDigit(ch);
	}

	public final static char CharAt(String Text, int Pos) {
		return Text.charAt(Pos);
	}

	public final static int ParseInt(String Text) {
		return Integer.parseInt(Text);
	}

	public final static Method LookupMethod(Object Callee, String MethodName) {
		if(MethodName != null) {
			// DebugP("looking up method : " + Callee.getClass().getSimpleName() + "." + MethodName);
			Method[] methods = Callee.getClass().getMethods();
			for(int i = 0; i < methods.length; i++) {
				if(MethodName.equals(methods[i].getName())) {
					return methods[i];
				}
			}
			DebugP("method not found: " + Callee.getClass().getSimpleName() + "." + MethodName);
		}
		return null; /*throw new GtParserException("method not found: " + callee.getClass().getName() + "." + methodName);*/
	}

	public final static boolean EqualsMethod(Method m1, Method m2) {
		if(m1 == null) {
			return (m2 == null) ? true : false;
		} else {
			return (m2 == null) ? false : m1.equals(m2);
		}
	}

	public final static TokenFunc CreateOrReuseTokenFunc(GtDelegateToken f, TokenFunc prev) {
		if(prev != null && EqualsMethod(prev.Func.Method, f.Method)) {
			return prev;
		}
		return new TokenFunc(f, prev);
	}

	public final static int ApplyTokenFunc(Object Self, Method Method, Object TokenContext, String Text, int pos) {
		try {
			Integer n = (Integer)Method.invoke(Self, TokenContext, Text, pos);
			return n.intValue();
		}
		catch (InvocationTargetException e) {
			e.printStackTrace();
		}
		catch (IllegalArgumentException e) {
			e.printStackTrace();
		} 
		catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		Exit(1, "Failed ApplyTokenFunc");
		return -1;
	}

	public final static SyntaxTree ApplyMatchFunc(Object Self, Method Method, Object Pattern, Object LeftTree, Object TokenContext) {
		try {
			return (SyntaxTree)Method.invoke(Self, Pattern, LeftTree, TokenContext);
		}
		catch (InvocationTargetException e) {
			e.printStackTrace();
		}
		catch (IllegalArgumentException e) {
			e.printStackTrace();
		} 
		catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		Exit(1, "Failed ApplyMatchFunc");
		return null;
	}

	public final static TypedNode ApplyTypeFunc(Object Self, Method Method, Object Gamma, Object ParsedTree, Object TypeInfo) {
		try {
			return (TypedNode)Method.invoke(Self, Gamma, ParsedTree, TypeInfo);
		}
		catch (InvocationTargetException e) {
			e.printStackTrace();
		}
		catch (IllegalArgumentException e) {
			e.printStackTrace();
		} 
		catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		Exit(1, "Failed ApplyTypeFunc");
		return null;
	}

	public final static GtType[] CompactTypeList(ArrayList<GtType> List) {
		GtType[] Tuple = new GtType[List.size()];
		for(int i = 0; i < List.size(); i++) {
			Tuple[i] = List.get(i);
		}
		return Tuple;
	}

	public final static String[] CompactStringList(ArrayList<String> List) {
		if(List == null) return null;
		String[] Tuple = new String[List.size()];
		for(int i = 0; i < List.size(); i++) {
			Tuple[i] = List.get(i);
		}
		return Tuple;
	}

	public final static GreenTeaGenerator CodeGenerator(String Option) {
		if(Option.equals("--js") || Option.equals("--javascript")) {
			return new JavaScriptSourceGenerator();
		}
		else if(Option.equals("--perl")) {
			return new CSourceGenerator();
		}
		else if(Option.equals("--c")) {
			return new CSourceGenerator();
		}
		return new JavaSourceGenerator();
	}
	
	public final static String LoadFile(String FileName) {
		File f = new File(FileName);
		byte[] b = new byte[(int) f.length()];
		FileInputStream fi;
		try {
			fi = new FileInputStream(f);
			fi.read(b);
			fi.close();
			return new String(b);
		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return null;
	}
}

File: src/TypeScript/LangDeps.js
Array.prototype.size = function () {
    return this.length;
};

Array.prototype.add = function (v) {
    this.push(v);
};

Array.prototype.get = function (i) {
    return this[i];
};

Array.prototype.set = function (i, v) {
    this[i] = v;
};

Array.prototype.remove = function (i) {
    var v = this[i];
    this.splice(i, 1);
    return v;
};

Object.prototype["equals"] = function (other) {
    return (this === other);
};

String.prototype["startsWith"] = function (key) {
    return this.lastIndexOf(key, 0) == 0;
};

String.prototype["replaceAll"] = function (key, rep) {
    this.replace(key, rep);
};

String.prototype["equals"] = function (other) {
    return (this == other);
};

var GtMap = (function () {
    function GtMap() {
        this.map = new Object();
        this.length = 0;
    }
    GtMap.prototype.get = function (index) {
        return this.map[index];
    };
    GtMap.prototype.put = function (key, obj) {
        this.length++;
        this.map[key] = obj;
    };
    GtMap.prototype.size = function () {
        return this.length;
    };
    return GtMap;
})();

var LangDeps = (function () {
    function LangDeps() {
    }
    LangDeps.Assert = function (expect) {
        if (!expect) {
            throw new Error();
        }
    };

    LangDeps.println = function (msg) {
        console.log(msg);
    };

    LangDeps.DebugP = function (msg) {
        if (DebugPrintOption) {
            LangDeps.println("DEBUG" + LangDeps.GetStackInfo(2) + ": " + msg);
        }
    };

    LangDeps.GetStackInfo = function (depth) {
        return " ";
    };

    LangDeps.IsWhitespace = function (ch) {
        return ch == 32 || ch == 9;
    };

    LangDeps.IsLetter = function (ch) {
        if (ch > 90) {
            ch -= 0x20;
        }
        return 65 <= ch && ch <= 90;
    };

    LangDeps.IsDigit = function (ch) {
        return 48 <= ch && ch <= 57;
    };

    LangDeps.CharAt = function (Text, Pos) {
        return Text.charCodeAt(Pos);
    };

    LangDeps.CharToString = function (code) {
        return String.fromCharCode(code);
    };

    LangDeps.ParseInt = function (Text) {
        return Text - 0;
    };

    LangDeps.LookupMethod = function (Callee, MethodName) {
        return Callee[MethodName];
    };

    LangDeps.EqualsMethod = function (m1, m2) {
        return m1 === m2;
    };

    LangDeps.CreateOrReuseTokenFunc = function (f, prev) {
        if (prev != null && LangDeps.EqualsMethod(prev.Func, f)) {
            return prev;
        }
        return new TokenFunc(f, prev);
    };

    LangDeps.ApplyTokenFunc = function (Delegate, TokenContext, Text, pos) {
        try  {
            return Delegate(TokenContext, Text, pos);
        } catch (e) {
            console.log(e);
        }
        return -1;
    };

    LangDeps.ApplyMatchFunc = function (Delegate, Pattern, LeftTree, TokenContext) {
        try  {
            return Delegate(Pattern, LeftTree, TokenContext);
        } catch (e) {
            console.log(e);
        }
        return null;
    };

    LangDeps.ApplyTypeFunc = function (Delegate, Gamma, ParsedTree, TypeInfo) {
        try  {
            return Delegate(Gamma, ParsedTree, TypeInfo);
        } catch (e) {
            console.log(e);
        }
        return null;
    };

    LangDeps.CompactTypeList = function (BaseIndex, List) {
        var Tuple = new Array(List.length - BaseIndex);
        for (var i = BaseIndex; i < List.length; i++) {
            Tuple[i] = List[i];
        }
        return Tuple;
    };

    LangDeps.CompactStringList = function (List) {
        if (List == null)
            return null;
        var Tuple = new Array(List.length);
        for (var i = 0; i < List.length; i++) {
            Tuple[i] = List[i];
        }
        return Tuple;
    };

    LangDeps.CodeGenerator = function (Option) {
        return new JavaScriptSourceGenerator();
    };

    LangDeps.HasFile = function (FileName) {
        throw new Error("LangDeps.LoadFile is not implemented for this environment");
        return "";
    };

    LangDeps.LoadFile = function (FileName) {
        throw new Error("LangDeps.LoadFile is not implemented for this environment");
        return "";
    };

    LangDeps.Usage = function () {
    };

    LangDeps.ReadLine = function (prompt) {
        return "";
    };
    return LangDeps;
})();
